### DISCUSSION TITLE
when use copy_current_request_context and ThreadPoolExecutor, I got error.

### DISCUSSION THREAD

**Discussion by Alvin4567** [pallets/flask#5254] [Q&A]
I want to use the main thread's request context in the sub-thread. So I use the decorator copy_current_request_context. But I got an error that **<Token var=<ContextVar name='flask.request_ctx' at 0x1067aef90> at 0x106c71280> was created in a different Context**. 

**The weirdest thing is when I debug the code one by one, I will get the expected results sometimes.**

This is a simple code. 

```
import random
import string
import requests

from flask import Flask, copy_current_request_context
from concurrent.futures import ThreadPoolExecutor

app = Flask(__name__)

def get_trace_id(number):
    trace_id =  "".join(random.sample(string.ascii_letters, number))
    print(requests.get("https://www.google.com").text)
    return trace_id


@app.route("/")
def hello():
    @copy_current_request_context
    def inner_func(number):
        return get_trace_id(number)

    with ThreadPoolExecutor(20) as executor:
        res_iter = executor.map(inner_func, [1, 2, 3, 4, 5])
    return list(res_iter)


if __name__ == '__main__':
    app.run()
```

Environment:

- Python version: 3.8.17
- Flask version: 2.3.3

**Comment by danderwald** [comment]
I am facing the same issue after upgrading flask to v 2.2.5 (python 3.11)

**Comment by danderwald** [comment]
Interesting.

not working:

```python
    @copy_current_request_context
    def do_something(param):
       pass

    loop.run_in_executor(
        executor,
        do_something
        param
    )
```

working:

```python
    def do_something(param):
       pass

    loop.run_in_executor(
        executor,
        copy_current_request_context(do_something)
        param
    )
```

after removing `copy_current_request_context` as a decorator and call it explicitly, I could not reproduce the behavior until now.

**Comment by zschneider** [comment]
I'm also encountering this issue, and neither approaches above worked for me. I'm on flask v 2.3.3 on Python 3.11. Were you able to get that to work consistently @danderwald?

**Comment by kasium** [comment]
Hi all,

I encountered the same issue and found the root cause, so I post it here for others also hitting it.
If you add logging to the constructor of RequestContext you see that only 2 are generated and not n+1 (where n is the number of threads).

Reason for that: copy_current_request_context is executed only once in the current thread and since copy_current_request_context creates the copy of the request context when the decorator is applied and not when the function is called, the same request context is used in all threads leading to the issue above. In order to fix it, you need to perform the copy in each worker thread. 

Saying this, copy_current_request_context  cannot be used on a function which is passed into a thread pool
